helix-rest/src/main/java/org/apache/helix/rest/client/CustomRestClientImpl.java

```diff
@@ -110,42 +152,102 @@ protected JsonNode getJsonObject(HttpResponse httpResponse) throws IOException {
     HttpPost postRequest = new HttpPost(url);
-    try {
-      postRequest.setHeader("Accept", ACCEPT_CONTENT_TYPE);
-      StringEntity entity = new StringEntity(OBJECT_MAPPER.writeValueAsString(payloads),
-          ContentType.APPLICATION_JSON);
-      postRequest.setEntity(entity);
-      LOG.info("Executing request: {}, headers: {}, entity: {}", postRequest.getRequestLine(),
-          postRequest.getAllHeaders(), postRequest.getEntity());
-
-      HttpResponse response = _httpClient.execute(postRequest);
-      int status = response.getStatusLine().getStatusCode();
-      if (status != HttpStatus.SC_OK) {
-        LOG.warn("Received non-200 status code: {}, payloads: {}", status, payloads);
+    int retries = 0;
+    HttpResponse response = null;
+
+    while (retries < MAX_REDIRECT) {
+      try {
+        postRequest.setHeader("Accept", ACCEPT_CONTENT_TYPE);
+        StringEntity entity = new StringEntity(OBJECT_MAPPER.writeValueAsString(payloads),
+            ContentType.APPLICATION_JSON);
+        postRequest.setEntity(entity);
+        LOG.info("Executing request: {}, headers: {}, entity: {}", postRequest.getRequestLine(),
+            postRequest.getAllHeaders(), postRequest.getEntity());
+
+        // Execute the request
+        response = _httpClient.execute(postRequest);
+        int status = response.getStatusLine().getStatusCode();
+
+        if (status == HttpStatus.SC_OK) {
+          return response;  // Return the successful response
+        } else if (status == HttpStatus.SC_MOVED_TEMPORARILY) {
+          // If receiving 302 Found (redirect), handle the redirection
+          Header locationHeader = response.getFirstHeader("Location");
+          if (locationHeader != null) {
+            String redirectUrl = locationHeader.getValue();
+            LOG.info("Redirecting to: {}", redirectUrl);
+            // Update the post request with the new URL
+            postRequest = new HttpPost(redirectUrl);
+            retries++;
+            LOG.info("Retrying redirect (attempt {} of {})", retries, MAX_REDIRECT);
+          } else {
+            LOG.warn("Received 302 but no Location header is present, stopping retries.");
+            break;  // Break out if there is no valid redirect location
+          }
+        } else {
+          LOG.warn("Received non-200 and non-302 status code: {}, payloads: {}", status, payloads);
+          return response;  // Return response without retry
+        }
+      } catch (IOException e) {
+        // Release connection to be reused and avoid connection leakage
+        postRequest.releaseConnection();
+        throw e;
       }
+    }
 
-      return response;
-    } catch (IOException e) {
-      // Release connection to be reused and avoid connection leakage.
-      postRequest.releaseConnection();
-      throw e;
+    if (retries >= MAX_REDIRECT) {
+      throw new IOException("Maximum redirect retries reached.");
```

Comment:
Why throw exp here?

Reply:
Ya, it's probably not needed. We can just return the response in that case and the method `handleResponse` will generate the proper exception.
Will remove it.